demos/http/http_demo_s3_upload/http_demo_s3_upload.c

```diff
@@ -0,0 +1,626 @@
+                            S3_PRESIGNED_PUT_URL,
+                            httpStatus ) );
+                returnStatus = EXIT_FAILURE;
+            }
+        }
+
+        /**************************** Connect. ******************************/
+
+        /* Establish TLS connection on top of TCP connection using OpenSSL. */
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            /* Attempt to connect to the HTTP server. If connection fails, retry after
+             * a timeout. Timeout value will be exponentially increased till the maximum
+             * attempts are reached or maximum timeout value is reached. The function
+             * returns EXIT_FAILURE if the TCP connection cannot be established to
+             * broker after configured number of attempts. */
+            returnStatus = connectToServerWithBackoffRetries( connectToServer,
+                                                              &networkContext );
+
+            if( returnStatus == EXIT_FAILURE )
+            {
+                /* Log error to indicate connection failure after all
+                 * reconnect attempts are over. */
+                LogError( ( "Failed to connect to HTTP server %s.",
+                            &serverHost ) );
+            }
+        }
+
+        /* Define the transport interface. */
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            ( void ) memset( &transportInterface, 0, sizeof( transportInterface ) );
+            transportInterface.recv = Openssl_Recv;
+            transportInterface.send = Openssl_Send;
+            transportInterface.pNetworkContext = &networkContext;
+        }
+
+        /********************** Upload S3 Object File. **********************/
+
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            returnStatus = uploadS3ObjectFile( &transportInterface,
+                                               &serverHost,
+                                               HTTP_METHOD_PUT,
+                                               pPath );
+        }
+
+        /******************* Verify S3 Object File Upload. ********************/
+
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            /* Retrieve the path location from S3_PRESIGNED_GET_URL. This function returns the length of the path
+             * without the query into pathLen. */
+            httpStatus = getUrlPath( S3_PRESIGNED_GET_URL,
+                                     strlen( S3_PRESIGNED_GET_URL ),
+                                     &pPath,
+                                     &pathLen );
+
+            if( httpStatus != HTTP_SUCCESS )
+            {
+                LogError( ( "An error occurred in getUrlPath() on URL %s. Error code: %d",
+                            S3_PRESIGNED_GET_URL,
+                            httpStatus ) );
+                returnStatus = EXIT_FAILURE;
+            }
```

Comment:
I think it would be best to leave the error logging up to the function responsible for doing that work. You can then use a ternary operator here to make it look cleaner.

Reply:
Sounds good -- moved all logging tasks to their respective functions.